broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,317 @@
+package backlog
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/log"
+	m "github.com/offchainlabs/nitro/broadcaster/message"
+	"github.com/offchainlabs/nitro/util/arbmath"
+)
+
+var (
+	errDropSegments       = errors.New("remove previous segments from backlog")
+	errSequenceNumberSeen = errors.New("sequence number already present in backlog")
+	errOutOfBounds        = errors.New("message not found in backlog")
+)
+
+type Backlog interface {
+	Head() BacklogSegment
+	Append(bm *m.BroadcastMessage) error
+	Get(start, end uint64) (*m.BroadcastMessage, error)
+	Count() int
+}
+
+type backlog struct {
+	head          atomic.Pointer[backlogSegment]
+	tail          atomic.Pointer[backlogSegment]
+	lookupLock    sync.RWMutex
+	lookupByIndex map[uint64]*atomic.Pointer[backlogSegment]
```

Comment:
Can lookupByIndex map to a backlogSegment directly instead of an atomic.Pointer to one?

Reply:
Good shout, changed it 👍